api/application.go

```diff
@@ -9,9 +11,49 @@ var EndpointApplications = Endpoint + "applications/"
 
 // https://discord.com/developers/docs/interactions/slash-commands#create-global-application-command-json-params
 type CreateCommandData struct {
-	Name        string                  `json:"name"`
-	Description string                  `json:"description"`
-	Options     []discord.CommandOption `json:"options"`
+	Name                string                  `json:"name"`
+	Description         string                  `json:"description"`
+	Options             []discord.CommandOption `json:"options"`
+	NoDefaultPermission bool                    `json:"-"`
+	Type                discord.CommandType     `json:"type,omitempty"`
+}
+
+func (c CreateCommandData) MarshalJSON() ([]byte, error) {
+	type RawCreateCommandData CreateCommandData
+	cmd := struct {
+		RawCreateCommandData
+		DefaultPermission bool `json:"default_permission"`
+	}{RawCreateCommandData: (RawCreateCommandData)(c)}
+
+	// Discord defaults default_permission to true, so we need to invert the
+	// meaning of the field (>No<DefaultPermission) to match Go's default
+	// value, false.
+	cmd.DefaultPermission = !c.NoDefaultPermission
+
+	return json.Marshal(cmd)
+}
+
+func (c *CreateCommandData) UnmarshalJSON(data []byte) error {
+	type RawCreateCommandData CreateCommandData
+	cmd := struct {
+		*RawCreateCommandData
+		DefaultPermission bool `json:"default_permission"`
+	}{RawCreateCommandData: (*RawCreateCommandData)(c)}
+	if err := json.Unmarshal(data, &cmd); err != nil {
+		return err
+	}
+
+	// Discord defaults default_permission to true, so we need to invert the
+	// meaning of the field (>No<DefaultPermission) to match Go's default
+	// value, false.
+	c.NoDefaultPermission = !cmd.DefaultPermission
+
+	// Discord defaults type to 1 if omitted.
+	if c.Type == discord.CommandType(0) {
```

Comment:
```suggestion
	if c.Type == 0 {
```

The explicit conversion of `0` to `discord.CommandType` is unnecessary, as Go can automatically infer the literal's type from `c.Type`.